### PR TITLE
refactor(contracts): Modernize BasePaymaster with custom errors and dedicated interface

### PR DESCRIPTION
**Pull request by adamgall**
Depends on https://github.com/decentdao/decent-contracts/pull/410

Fixes [ENG-1237](https://linear.app/decent-labs/issue/ENG-1237/modernize-basepaymaster-with-custom-errors-and-dedicated-interface)

Comprehensively refactored BasePaymaster to align with modern Solidity patterns, creating IBasePaymaster interface for its unique public API and replacing all require statements with gas-efficient custom errors.

Changes:
- Created IBasePaymaster interface with custom errors and unique public functions
- Updated BasePaymaster to inherit from IBasePaymaster
- Replaced all require statements and string reverts with custom errors
- Reorganized contract structure to follow codebase patterns
- Created comprehensive unit tests for BasePaymaster (25 tests)
- Added ConcreteBasePaymaster mock for testing abstract contract functionality
- Updated DecentPaymasterV1 to support IBasePaymaster interface in ERC165
- Converted all NatSpec documentation from /// to /** */ format
- Added NatSpec documentation to all constants and functions
- Added override modifiers to all non-internal functions
- Added entryPoint() function to IBasePaymaster interface
- Enhanced test coverage with EntryPoint impersonation and edge cases

# Project Plan

## Overview

Refactor the BasePaymaster abstract contract to replace all `require` statements with custom errors, aligning it with the modern Solidity patterns used throughout the Decent contracts codebase. This includes creating appropriate unit tests and a concrete mock implementation for testing.

## Implementation Steps (Completed)

1. **Create IBasePaymaster Interface** ✓

   - Created `IBasePaymaster.sol` in `contracts/interfaces/decent/deployables/`
   - Defined custom errors in the interface
   - Defined the unique public functions that BasePaymaster exposes (entryPoint, deposit, withdrawTo, addStake, getDeposit, unlockStake, withdrawStake)
   - Added comprehensive NatSpec documentation for each function
   - Updated BasePaymaster to inherit from IBasePaymaster

2. **Update BasePaymaster Contract** ✓

   - Added IBasePaymaster to the inheritance list
   - Removed error definitions from contract (now in interface)
   - Updated function documentation to use @inheritdoc for interface functions
   - Reorganized contract structure to follow codebase patterns (STATE VARIABLES → CONSTRUCTOR & INITIALIZERS → IPaymaster → INTERNAL HELPERS)
   - Updated parameter names to use trailing underscore convention

3. **Refactor Require Statements** ✓

   - Replaced `require` in `_validateEntryPointInterface()` with `InvalidEntryPointInterface`
   - Replaced `require` in `_requireFromEntryPoint()` with `CallerNotEntryPoint`
   - Replaced `revert("must override")` in `_postOp()` with `PostOpNotImplemented`

4. **Update DecentPaymasterV1** ✓

   - Added IBasePaymaster import
   - Updated supportsInterface function to include IBasePaymaster
   - Updated NatSpec comment to mention IBasePaymaster support
   - Added IBasePaymaster\_\_factory to the test imports and supportsInterface test

5. **Create Mock Implementation for Testing** ✓

   - Created `ConcreteBasePaymaster.sol` in `contracts/mocks/concretes/deployables/account-abstraction/`
   - Implemented abstract methods: `_validatePaymasterUserOp` and `_postOp`
   - Added test helper functions for configuring validation behavior
   - Added test helpers to directly call internal functions for testing

6. **Create Unit Test File** ✓

   - Created `BasePaymaster.test.ts` in `test/unit/deployables/account-abstraction/`
   - Tested all public/external functions (deposit, withdrawTo, addStake, etc.)
   - Tested all custom error scenarios including PostOpNotImplemented
   - Tested access control (onlyOwner modifiers)
   - Tested EntryPoint validation and interactions
   - Added EntryPoint impersonation tests for successful validation flows
   - Added comprehensive edge case testing for deposits and withdrawals
   - Tested context and validation data handling in \_validatePaymasterUserOp
   - All 25 tests passing with comprehensive code coverage

7. **Update Existing Tests** ✓

   - Searched for tests expecting old revert messages - found none in test files
   - Added test for IBasePaymaster interface support in DecentPaymasterV1.test.ts
   - All existing tests continue to pass (35 tests in DecentPaymasterV1)

8. **Code Quality Checks** ✓
   - Ran prettier formatting: `npm run pretty` - all files unchanged
   - Ran TypeScript linting: `npm run lint` - pre-existing import/extensions errors unrelated to changes
   - Ran Solidity linting: `npm run solhint:check` - passed without issues
   - All new code follows project conventions and standards

## IBasePaymaster Interface Structure

The interface includes both custom errors and the unique public functions:

```solidity
// SPDX-License-Identifier: AGPL-3.0
pragma solidity ^0.8.30;

/**
 * @title IBasePaymaster
 * @notice Interface for base paymaster functionality in ERC-4337
 * @dev Extends the standard paymaster with deposit and staking management.
 * This interface defines functions that are common to paymaster implementations
 * but not part of the core IPaymaster interface from the account-abstraction package.
 */
interface IBasePaymaster {
  // --- Errors ---

  /** @notice Thrown when the provided address does not implement IEntryPoint interface */
  error InvalidEntryPointInterface();

  /** @notice Thrown when caller is not the configured EntryPoint */
  error CallerNotEntryPoint();

  /** @notice Thrown when _postOp is called but not overridden by the inheriting contract */
  error PostOpNotImplemented();

  // --- View Functions ---

  /**
   * @notice Returns the EntryPoint contract address
   * @return entryPoint The EntryPoint contract instance
   */
  function entryPoint() external view returns (IEntryPoint entryPoint);

  /**
   * @notice Returns the current deposit balance for this paymaster
   * @return depositBalance The amount of ETH deposited for gas payments
   */
  function getDeposit() external view returns (uint256 depositBalance);

  // --- State-Changing Functions ---

  /**
   * @notice Adds to the deposit balance for paying gas fees
   * @dev Deposits ETH to the EntryPoint for this paymaster to use
   */
  function deposit() external payable;

  /**
   * @notice Withdraws from the deposit balance
   * @param withdrawAddress_ The address to receive the withdrawn funds
   * @param amount_ The amount of ETH to withdraw
   */
  function withdrawTo(address payable withdrawAddress_, uint256 amount_) external;

  /**
   * @notice Stakes tokens with the EntryPoint
   * @param unstakeDelaySec_ The delay in seconds before stake can be withdrawn
   */
  function addStake(uint32 unstakeDelaySec_) external payable;

  /**
   * @notice Unlocks the stake for withdrawal
   * @dev Must wait for unstakeDelay before withdrawing
   */
  function unlockStake() external;

  /**
   * @notice Withdraws the unlocked stake
   * @param withdrawAddress_ The address to receive the withdrawn stake
   */
  function withdrawStake(address payable withdrawAddress_) external;
}
```

## Benefits

- **Gas Efficiency**: Custom errors use less gas than string-based require messages
- **Better Developer Experience**: More descriptive and type-safe error handling
- **Consistency**: Aligns with the error handling patterns in the rest of the codebase
- **Maintainability**: Easier to track and handle specific error conditions
- **Architectural Clarity**: IBasePaymaster interface provides clear documentation for paymaster-specific functions
- **Extensibility**: Provides a proper interface for the unique functions BasePaymaster adds beyond IPaymaster

## Notes

- BasePaymaster has already diverged significantly from the original @account-abstraction package version
- The contract already uses EIP-7201 namespaced storage and initializer patterns
- This refactoring completes the modernization of the contract
- The mock implementation will be useful for testing other paymaster features in isolation
- BasePaymaster does introduce new public functions (deposit, withdrawTo, addStake, etc.) that warrant a dedicated interface
- DecentPaymasterV1 properly includes IBasePaymaster in its supportsInterface implementation

### CODE REVIEW

File: contracts/deployables/account-abstraction/BasePaymaster.sol [hunk 1]
```diff
@@ -90,152 +118,170 @@ abstract contract BasePaymaster is IPaymaster, OwnableUpgradeable {
...
+     * @param actualUserOpFeePerGas_ The gas price paid by this UserOp
+     * @custom:throws PostOpNotImplemented if not overridden by inheriting contract
+     */
+    function _postOp(
+        PostOpMode mode_,
+        bytes calldata context_,
+        uint256 actualGasCost_,
+        uint256 actualUserOpFeePerGas_
+    ) internal virtual {
+        (mode_, context_, actualGasCost_, actualUserOpFeePerGas_); // unused params
```
Comment (tbwebb22) [inline contracts/deployables/account-abstraction/BasePaymaster.sol:253]: what is the purpose of this line?

Comment (adamgall) [reply contracts/deployables/account-abstraction/BasePaymaster.sol:253]: to satisfy the compiler warnings about unused params. but yeah we could also just remove the names of those params from the function signature.

Comment (tbwebb22) [reply contracts/deployables/account-abstraction/BasePaymaster.sol:253]: yeah that would be my preference. Your call though. I'll approve as everything else looked good

**Review by tbwebb22 — APPROVED** [review]

File: contracts/deployables/account-abstraction/BasePaymaster.sol (hunk 1 above)
Comment (adamgall) [reply contracts/deployables/account-abstraction/BasePaymaster.sol:253]: so after playing around with this, i'm going to leave it as is. If we remove the param names from the function signature, then the NatSpec docs become errors unless I delete those `@param` lines. And considering the point of this function is simply to say "hey you should override this in your concrete instance", i think it's important for the params documentation to exist.
